nvbench/detail/measure_cupti.cuh

```diff
@@ -0,0 +1,158 @@
+/*
+ *  Copyright 2021 NVIDIA Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 with the LLVM exception
+ *  (the "License"); you may not use this file except in compliance with
+ *  the License.
+ *
+ *  You may obtain a copy of the License at
+ *
+ *      http://llvm.org/foundation/relicensing/LICENSE.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#pragma once
+
+#include <nvbench/config.cuh>
+
+#ifdef NVBENCH_HAS_CUPTI
+#include <nvbench/blocking_kernel.cuh>
+#include <nvbench/cpu_timer.cuh>
+#include <nvbench/cuda_call.cuh>
+#include <nvbench/cuda_timer.cuh>
+#include <nvbench/cupti_profiler.cuh>
+#include <nvbench/device_info.cuh>
+#include <nvbench/exec_tag.cuh>
+#include <nvbench/launch.cuh>
+
+#include <nvbench/detail/kernel_launcher_timer_wrapper.cuh>
+#include <nvbench/detail/l2flush.cuh>
+#include <nvbench/detail/statistics.cuh>
+
+#include <cuda_runtime.h>
+
+#include <algorithm>
+#include <utility>
+#include <vector>
+
+namespace nvbench
+{
+
+struct state;
+
+namespace detail
+{
+
+// non-templated code goes here:
+struct measure_cupti_base
+{
+  explicit measure_cupti_base(nvbench::state &exec_state);
+  measure_cupti_base(const measure_cupti_base &) = delete;
+  measure_cupti_base(measure_cupti_base &&)      = delete;
+  measure_cupti_base &operator=(const measure_cupti_base &) = delete;
+  measure_cupti_base &operator=(measure_cupti_base &&) = delete;
+
+protected:
+  struct kernel_launch_timer;
+
+  void check();
+  void generate_summaries();
+
+  __forceinline__ void flush_device_l2()
+  {
+    m_l2flush.flush(m_launch.get_stream());
+  }
+
+  __forceinline__ void sync_stream() const
+  {
+    NVBENCH_CUDA_CALL(cudaStreamSynchronize(m_launch.get_stream()));
+  }
+
+  nvbench::state &m_state;
+
+  nvbench::launch m_launch;
+  nvbench::detail::l2flush m_l2flush;
+
+  cupti_profiler m_cupti;
+};
+
+struct measure_cupti_base::kernel_launch_timer
+{
+  explicit kernel_launch_timer(measure_cupti_base &measure)
+      : m_measure{measure}
+  {}
+
+  __forceinline__ void start()
+  {
+#ifdef NVBENCH_HAS_CUPTI
```

Comment:
This is redundant with line 23.

I think we can probably remove all of these checks and just ifdef out the `#include <nvbench/detail/measure_cupti.cuh>` in `state_exec.cuh`.